src/storage/persistence.rs

```diff
@@ -0,0 +1,73 @@
+use crate::storage::{
+    storage_entry::{storage_entry, StorageEntry},
+    SessionStorage,
+};
+use dioxus::prelude::*;
+use dioxus_signals::use_signal;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+use super::storage_entry::UseStorageEntry;
+
+/// A persistent storage hook that can be used to store data across application reloads.
+///
+/// Depending on the platform this uses either local storage or a file storage
+#[allow(clippy::needless_return)]
+pub fn use_persistent<T: Serialize + DeserializeOwned + Default + Clone + 'static>(
+    cx: &ScopeState,
+    key: impl ToString,
+    init: impl FnOnce() -> T,
+) -> &UseStorageEntry<SessionStorage, T> {
+    let mut init = Some(init);
+    let state = {
+        #[cfg(feature = "ssr")]
+        {
+            use_ref(cx, || {
+                StorageEntry::<SessionStorage, T>::new(key.to_string(), init.take().unwrap()(), None)
+            })
+        }
+        #[cfg(all(not(feature = "ssr"), not(feature = "hydrate")))]
+        {
+            use_signal(cx, || {
```

Comment:
We can use signal here, but then we need to change the other flags to also use signals instead of use_ref. Currently the ssr feature is broken

Reply:
yeah I focused on the one case I was currently testing for the prototype, will swap the rest out

Reply:
How does SSR benefit from this feature? Curious if it's even worth implementing. Maybe we should just panic for SSR to avoid devs thinking it's doing something it's not. I could see this being useful for fullstack though (store the settings in the browser and communicate them back to the backend if needed)

Reply:
> How does SSR benefit from this feature? Curious if it's even worth implementing. Maybe we should just panic for SSR to avoid devs thinking it's doing something it's not. I could see this being useful for fullstack though (store the settings in the browser and communicate them back to the backend if needed)

SSR doesn't really use this feature, but we still need to provide a default value instead of panicing in SSR mode to make hydration work.

The approach I used initially did this:
On the server: Always use the value that should appear when storage is empty
On the client: First hydrate with the value that should appear when storage is empty on the first renderer, then update to the real value after hydration is finished